File: web/lib/computation-contract.ts
/**
 * Computed answers outside a chat turn: listed, compared, continued and
 * refreshed. Every figure and every difference is computed by the backend;
 * the web only formats what it receives.
 */
import type { DecisionComputation, DecisionRerun } from "./decision-contract";

type ComputationText = { locale_key: string; interpolation_args: Record<string, string | number | boolean | null> };

export type ComputedAnswerRef = { conversation_id: string; message_id: string };
export type ComputedAnswerSummary = {
  conversation_id: string; message_id: string; kind: string; asked: string | null; computed_at: string; symbols: string[];
};
export type ComparedAnswer = {
  conversation_id: string; message_id: string; asked: string | null; computed_at: string; card: Record<string, unknown>;
};
export type ComputationDifference = {
  section: "answer" | "row" | "input"; name: string; label: ComputationText; unit: ComputationText | null;
  left: number; right: number; difference: number;
};
export type ComputationComparison = { kind: string; left: ComparedAnswer; right: ComparedAnswer; differences: ComputationDifference[] };
export type ContinuedResult = { conversation: { id: string; title: string }; message_id: string };
export type ComputationRefresh = {
  computation: DecisionComputation; status: "refreshed" | "inputs_not_found"; rerun: DecisionRerun | null; sources: Record<string, unknown>[];
};
export type ContinuedFrom = { conversationId: string; messageId: string };

/** The source of a result continued in a new chat, read from backend metadata only. */
export function continuedFromMetadata(metadata: Record<string, unknown> | null | undefined): ContinuedFrom | null {
  const raw = metadata?.continued_from;
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) return null;
  const { conversation_id: conversationId, message_id: messageId } = raw as Record<string, unknown>;
  return typeof conversationId === "string" && conversationId && typeof messageId === "string" && messageId
    ? { conversationId, messageId }
    : null;
}

File: src/argus/api/computation_contract.py
"""Computed answers outside a chat turn: listed, compared, continued, refreshed.

Comparing and continuing are free and call no model or provider. Refreshing
the cited inputs is the one paid action; the user starts it and reviews the
result beside the stored answer, which is never rewritten.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from argus.api.decision_contract import DecisionComputation, DecisionRerun
from argus.api.schemas import Conversation
from argus.domain.tool_contracts import LocalizedText

MAX_LISTED_ANSWERS = 10


class ComputedAnswerRef(BaseModel):
    model_config = ConfigDict(extra="forbid")

    conversation_id: str = Field(min_length=1, max_length=64)
    message_id: str = Field(min_length=1, max_length=64)


class ComputedAnswerSummary(BaseModel):
    """One owned computed answer a comparison can pick."""

    conversation_id: str
    message_id: str
    kind: str = Field(max_length=80)
    asked: str | None = Field(default=None, max_length=500)
    computed_at: datetime
    symbols: list[str] = Field(default_factory=list, max_length=5)


class ComputedAnswerList(BaseModel):
    items: list[ComputedAnswerSummary] = Field(max_length=MAX_LISTED_ANSWERS)


class ComputationCompareRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    left: ComputedAnswerRef
    right: ComputedAnswerRef


class ComparedAnswer(BaseModel):
    conversation_id: str
    message_id: str
    asked: str | None = Field(default=None, max_length=500)
    computed_at: datetime
    card: dict[str, Any]


class ComputationDifference(BaseModel):
    """Right minus left for one fact both cards state as a number in one unit."""

    section: Literal["answer", "row", "input"]
    name: str
    label: LocalizedText
    unit: LocalizedText | None = None
    left: float
    right: float
    difference: float


class ComputationComparison(BaseModel):
    kind: str
    left: ComparedAnswer
    right: ComparedAnswer
    differences: list[ComputationDifference]


class ContinuedResultResponse(BaseModel):
    """The new chat and the one message that carries the continued result."""

    conversation: Conversation
    message_id: str


ComputationRefreshStatus = Literal["refreshed", "inputs_not_found"]


class ComputationRefreshResponse(BaseModel):
    """Freshly retrieved inputs computed beside the stored answer.

    ``refreshed`` carries one recomputed card per calculation in ``reruns``, in
    the computation's order; ``inputs_not_found`` means the retrieval named no
    cited input, so nothing was recomputed. ``sources`` lists the pages the
    retrieval read, dated.
    """

    computation: DecisionComputation
    status: ComputationRefreshStatus
    reruns: list[DecisionRerun] = Field(default_factory=list, max_length=4)
    sources: list[dict[str, Any]] = Field(default_factory=list)
